das/daser.go

```diff
@@ -42,28 +62,49 @@ func (d *DASer) Start(context.Context) error {
 		return err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	go d.sampling(ctx, sub)
+	// load latest DASed checkpoint
+	checkpoint, err := loadCheckpoint(d.cstore)
+	if err != nil {
+		return err
+	}
+	log.Infow("loaded checkpoint", "height", checkpoint)
+
+	dasCtx, cancel := context.WithCancel(context.Background())
 	d.cancel = cancel
+	d.sampleDn, d.catchUpDn = make(chan struct{}), make(chan struct{})
+
+	go d.sample(dasCtx, sub, checkpoint)
 	return nil
 }
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
 	d.cancel()
-	select {
-	case <-d.done:
-		d.cancel = nil
-		return nil
-	case <-ctx.Done():
-		return ctx.Err()
+	// wait for both sampling routines to exit
+	for i := 0; i < 2; i++ {
+		select {
+		case <-d.catchUpDn:
+			d.catchUpDn = nil
+		case <-d.sampleDn:
+			d.sampleDn = nil
+		case <-ctx.Done():
+			return ctx.Err()
+		}
 	}
+	d.cancel = nil
+	return nil
 }
 
-// sampling validates availability for each Header received from header subscription.
-func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
-	defer sub.Cancel()
-	defer close(d.done)
+// sample validates availability for each Header received from header subscription.
+func (d *DASer) sample(ctx context.Context, sub header.Subscription, checkpoint int64) {
+	defer func() {
+		sub.Cancel()
+		close(d.sampleDn)
+	}()
+
+	// kick off catchUpScheduler
+	go d.catchUpScheduler(ctx, checkpoint)
```

Comment:
Sample does not control the lifecycle of this routine, so pls move it to the constructor

Reply:
https://github.com/celestiaorg/celestia-node/pull/473/commits/a999cf1bf646fdd672732de603ac363256db3af4

Resolved?

Reply:
resolved?